This commit enables MarkDuplicate to process query-sorted input.

yfarjoun opened the pull request:
Additionally, if the input is query-sorted it will mark the supplementary and secondary reads in the same manner as their primary counter-parts (with the same query-name). Furthermore, in this case a unmapped read will be duplicate-marked as its mapped mate (assuming, it is mated of-course).

Some tests had to be changed slightly to accommodate for the different behaviors. Tests have been added to test for the expected functionality around supplementary and secondary reads (which was untested), in both query and coordinate-sorted inputs.

I think that it should be possible (and correct) to mark unmapped-records the same as their mapped mates even in coordinate-sorted inputs, but I will leave that for a future discussion.

**NOTE:**
This commit includes an API change regarding the CLP parameter ASSUME_SORTED. This boolean has been replaced with a SortOrder parameter name ASSUME_SORT_ORDER which defaults to null (making the default behavior the same). It enables the user to force either query-sort or coordinate-sort behavior.

nh13 commented on src/java/picard/sam/DuplicationMetrics.java:
```diff
@@ -65,6 +68,7 @@
     /** The estimated number of unique molecules in the library based on PE duplication. */
     public Long ESTIMATED_LIBRARY_SIZE;
 
+
```
:/

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -79,7 +80,12 @@
-                    "                 O=marked_duplicates.bam M=marked_dup_metrics.txt\n";
+                    "                 O=marked_duplicates.bam M=marked_dup_metrics.txt\n" +
+                    "\n" +
+                    "The program can take either coordinate-sorted or coordinate-sorted input, however the behaviour is slightly different." +
```
`behaviour` -> `behavior`.

geoffjentry replied:
:us:

yfarjoun replied:
I make this (specific) spelling "error" all the time..I guess I learnt :gb: back in Israel...

nh13 replied:
It's ok, when in **||**:maple_leaf:**||**, I would do the same.

eitanbanks replied:
As proof to how pedantic we've all become: you noticed that behavior was spelled wrong but didn't think about the actual meaning of the sentence:
"The program can take either coordinate-sorted or coordinate-sorted..."
I think there's a mistake in there...  :)

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -142,7 +148,10 @@
+    private SAMFileHeader.SortOrder sortOrder;
+
     private int numDuplicateIndices = 0;
+    private final long NO_SUCH_INDEX = Long.MAX_VALUE; // needs to be large so that thee >= test fails for query-sorted traversal
```
`thee`?

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -198,8 +207,17 @@ protected int doWork() {
-        outputHeader.setSortOrder(SAMFileHeader.SortOrder.coordinate);
-        for (final String comment : COMMENT) outputHeader.addComment(comment);
+
+        sortOrder = outputHeader.getSortOrder();
```
perhaps clearer if it was `sortOrder = header.getSortOrder();` and moved up.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -198,8 +207,17 @@ protected int doWork() {
+        log.info("Reads are assumed to be ordered by: " + sortOrder);
+
+        if (sortOrder != SAMFileHeader.SortOrder.coordinate && sortOrder != SAMFileHeader.SortOrder.queryname) {
+            throw new PicardException("This program requires input that are coordinate or query sorted.");
```
either or

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -198,8 +207,17 @@ protected int doWork() {
+            throw new PicardException("This program requires input that are coordinate or query sorted.");
+        }
+
+        //outputHeader.setSortOrder(SAMFileHeader.SortOrder.coordinate);
```
remove comment

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -229,14 +250,31 @@ protected int doWork() {
                 }
 
+                // Now try and figure out the next duplicate index (if going by coordinate. if going by query name, only do this
+                // if the query name has changed.
```
Missing `)` in this comment.  Also capitalize the second `if`.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -229,14 +250,31 @@ protected int doWork() {
+                // if the query name has changed.
+                final boolean needNextDuplicateIndex = recordInFileIndex > nextDuplicateIndex && //passed next duplicate
+                        ( sortOrder == SAMFileHeader.SortOrder.coordinate ||  //in coordinate sorted, always
+                                !rec.getReadName().equals(duplicateSetQueryName) && sortOrder == SAMFileHeader.SortOrder.queryname); //in querysorted if the query name has changed
```
Please put parentheses when you have `( A || B && C)` to explicitly make it either `( A || (B && C))` or `((A || B) && C)`.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -229,14 +250,31 @@ protected int doWork() {
+
+                final boolean isDuplicate = sortOrder == SAMFileHeader.SortOrder.queryname &&
+                        recordInFileIndex > nextDuplicateIndex &&
+                        rec.getReadName().equals(duplicateSetQueryName) ||
```
again, please don't mix `&&` and `||` without parentheses.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -229,14 +250,31 @@ protected int doWork() {
+                    nextDuplicateIndex = (this.duplicateIndexes.hasNext() ? this.duplicateIndexes.next() : NO_SUCH_INDEX);
+                }
+
+                final boolean isDuplicate = sortOrder == SAMFileHeader.SortOrder.queryname &&
```
can you make the order of each conditional the same as the order on line 263?

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -246,25 +284,39 @@ protected int doWork() {
-                    }
                 } else {
                     rec.setDuplicateReadFlag(false);
+                    //duplicateSetQueryName = null;
```
remove

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -246,25 +284,39 @@ protected int doWork() {
-            if (isOpticalDuplicate) nextOpticalDuplicateIndex = this.opticalDuplicateIndexes.hasNext() ? this.opticalDuplicateIndexes.next() : -1;
+            final boolean needNextOpticalDuplicateIndex = recordInFileIndex > nextOpticalDuplicateIndex && //passed next duplicate
+                    ( sortOrder == SAMFileHeader.SortOrder.coordinate ||  //in coordeinate sorted, always
+                            !rec.getReadName().equals(opticalDuplicateSetQueryName) && sortOrder == SAMFileHeader.SortOrder.queryname); //in querysorted if the query name has changed
```
ditto

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -246,25 +284,39 @@ protected int doWork() {
+            // if the query name has changed)
+            if (recordInFileIndex > nextOpticalDuplicateIndex && (
+                    !rec.getReadName().equals(opticalDuplicateSetQueryName) ||
+                            sortOrder == SAMFileHeader.SortOrder.coordinate)) {
```
ditto

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -246,25 +284,39 @@ protected int doWork() {
+            final boolean isOpticalDuplicate = sortOrder == SAMFileHeader.SortOrder.queryname &&
+                    recordInFileIndex > nextOpticalDuplicateIndex &&
+                    rec.getReadName().equals(opticalDuplicateSetQueryName) ||
+                    recordInFileIndex == nextOpticalDuplicateIndex;
```
ditto

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -373,23 +428,34 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
                 pgIdsSeen.add(rec.getStringAttribute(SAMTag.PG.name()));
             }
 
+            //if working in querysorted, need to keep index of first record with any given query name.
```
space after `//` here and elsewhere?

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -373,23 +428,34 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
+                duplicateSetIndex = index;
+            }
+
+
```
you know I don't like these.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -359,6 +412,8 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
             this.libraryIdGenerator = new LibraryIdGenerator(header);
         }
 
+        String duplicateSetName = null;
```
I like `duplicateSetQueryName` better.  Perhaps even `duplicateQueryName` since duplicate sets aren't mentioned anywhere else in this code.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -373,23 +428,34 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
+            //if working in querysorted, need to keep index of first record with any given query name.
+            if(assumedSortOrder == SAMFileHeader.SortOrder.queryname && !rec.getReadName().equals(duplicateSetName)) {
+                duplicateSetName = rec.getReadName();
+                duplicateSetIndex = index;
```
consider this:
```
duplicateSetName  = rec.getReadName();
duplicateSetIndex = index;
```

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -373,23 +428,34 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
-                if (rec.getReferenceIndex() == -1) {
-                    // When we hit the unmapped reads with no coordinate, no reason to continue.
+                if (rec.getReferenceIndex() == -1 && assumedSortOrder == SAMFileHeader.SortOrder.coordinate) {
+                    // When we hit the unmapped reads with no coordinate, no reason to continue. (in coordinate sort)
```
parentheses before period?

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -373,23 +428,34 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
                     break;
                 }
                 // If this read is unmapped but sorted with the mapped reads, just skip it.
+
```
?

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -373,23 +428,34 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
 
                 if (rec.getReadPairedFlag() && !rec.getMateUnmappedFlag()) {
                     final String key = rec.getAttribute(ReservedTagConstants.READ_GROUP_ID) + ":" + rec.getReadName();
+
```
?

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -521,16 +587,12 @@ private void generateDuplicateIndexes(final boolean useBarcodes, final boolean i
                 if (nextChunk.size() > 1) {
                     markDuplicatePairs(nextChunk);
                 }
-
```
good work!

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -521,16 +587,12 @@ private void generateDuplicateIndexes(final boolean useBarcodes, final boolean i
         // First just do the pairs
         log.info("Traversing read pair information and detecting duplicates.");
         for (final ReadEndsForMarkDuplicates next : this.pairSort) {
-            if (firstOfNextChunk == null) {
```
Can you set `firstOfNextChunk = null` prior to line 610 (`for (final ReadEndsForMarkDuplicates next : this.fragSort)`).  I am trying to debate why `firstOfNextChunk` wouldn't need to be reset to `null` before starting on the fragments.

yfarjoun replied:
I think that the reason is that it will equal the firstOfNextChunk for the last chunk of pairs, which should be different than the first fragment...there might be an edge case...so I'll add the nullify.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -626,7 +687,10 @@ private void markDuplicatePairs(final List<ReadEndsForMarkDuplicates> list) {
                 addIndexAsDuplicate(end.read1IndexInFile);
-                addIndexAsDuplicate(end.read2IndexInFile);
+
+                //in query-sorted case, these will be the same.
```
space after `//`. so stingy.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicatesWithMateCigar.java:
```diff
@@ -120,7 +121,11 @@ protected int doWork() {
+            throw new PicardException("This program requires inputs in coordinate SortOrder");
+        }
+
+        //outputHeader.setSortOrder(SAMFileHeader.SortOrder.coordinate);
```
remove

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicatesWithMateCigar.java:
```diff
@@ -120,7 +121,11 @@ protected int doWork() {
         // Create the output header
         final SAMFileHeader outputHeader = header.clone();
-        outputHeader.setSortOrder(SAMFileHeader.SortOrder.coordinate);
+        if(outputHeader.getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
```
space after `if`

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicatesWithMateCigar.java:
```diff
@@ -24,6 +24,7 @@
 
 package picard.sam.markduplicates;
 
+import picard.PicardException;
```
this class is yours now

yfarjoun replied:
not sure what is meant here..

nh13 replied:
*WithMateCigar is yours now.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicatesWithMateCigar.java:
```diff
@@ -120,7 +121,11 @@ protected int doWork() {
+        }
+
+        //outputHeader.setSortOrder(SAMFileHeader.SortOrder.coordinate);
         for (final String comment : COMMENT) outputHeader.addComment(comment);
```
No love here?  `COMMENT.forEach(outputHeader::addComment);`

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -626,7 +687,10 @@ private void markDuplicatePairs(final List<ReadEndsForMarkDuplicates> list) {
-                addIndexAsDuplicate(end.read2IndexInFile);
+
+                //in query-sorted case, these will be the same.
+                // TODO: also in coordinate sorted, when one read is unmapped
```
TODO!

yfarjoun replied:
Yes...well, I wasn't sure if the change is appropriate for this PR. Do you think it matters? I can just as easily change the behavior for coordinate-sorted input...

nh13 commented on src/java/picard/sam/markduplicates/util/AbstractMarkDuplicatesCommandLineProgram.java:
```diff
@@ -74,12 +73,11 @@
     @Option(shortName = StandardOptionDefinitions.ASSUME_SORTED_SHORT_NAME,
-            doc = "If true, assume that the input file is coordinate sorted even if the header says otherwise.")
-    public boolean ASSUME_SORTED = false;
+            doc = "If not null, assume that the input file has this order even if the header says otherwise.",optional = true)
```
`,optional` -> `, optional`

nh13 commented on src/java/picard/sam/markduplicates/util/AbstractMarkDuplicatesCommandLineProgram.java:
```diff
@@ -74,12 +73,11 @@
-            doc = "If true, assume that the input file is coordinate sorted even if the header says otherwise.")
-    public boolean ASSUME_SORTED = false;
+            doc = "If not null, assume that the input file has this order even if the header says otherwise.",optional = true)
+    public SAMFileHeader.SortOrder ASSUME_ORDER = null;
```
`ASSUME_SORT_ORDER`?

nh13 commented on src/java/picard/sam/markduplicates/util/AbstractMarkDuplicatesCommandLineProgram.java:
```diff
@@ -205,19 +203,20 @@ protected SamHeaderAndIterator openInputs() {
             readers.add(reader);
         }
 
+        if(ASSUME_ORDER != null) {
```
space after `if`

nh13 commented on src/tests/java/picard/sam/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java:
```diff
@@ -540,4 +549,32 @@ public void testOpticalDuplicateClustersAddingSecondEndFirstSameCoordinate() {
+
+    @DataProvider(name = "TrueFalseSquared")
+    public Object[][] trueFalseSquaredData() {
+        List<Boolean> truth = CollectionUtil.makeList(true, false);
```
final

nh13 commented on src/tests/java/picard/sam/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java:
```diff
@@ -540,4 +549,32 @@ public void testOpticalDuplicateClustersAddingSecondEndFirstSameCoordinate() {
+    @DataProvider(name = "TrueFalseSquared")
+    public Object[][] trueFalseSquaredData() {
+        List<Boolean> truth = CollectionUtil.makeList(true, false);
+        return truth.stream().flatMap(pred1->truth.stream().map(pred2->new Boolean[]{pred1, pred2})).filter(o->o[0]||o[1]).toArray(Object[][]::new);
```
Oh wow, why not just this:
``` 
return new Object[][] {
  { true, true }, 
  { true, false},
  { false, true},
  { false, false}
};
```
?

yfarjoun replied:
'cause I was playing...sorry.

nh13 commented on src/tests/java/picard/sam/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java:
```diff
@@ -540,4 +549,32 @@ public void testOpticalDuplicateClustersAddingSecondEndFirstSameCoordinate() {
+        tester.addMappedFragment( "RUNID:1:1:15993:13361", 1, 400, markSecondaryAndSupplementaryRecordsLikeTheCannonical(), null, null, additionalFragSecondary, additionalFragSupplementary, DEFAULT_BASE_QUALITY);
+        tester.addMatePair("RUNID:1:1:15993:13361", 2, 41212324, 41212310, false, false, false, false, "33S35M", "19S49M", true, true, false, false, false, DEFAULT_BASE_QUALITY);
+        tester.addMatePair("RUNID:1:1:16020:13352", 2, 41212324, 41212319, false, false, true, true, "33S35M", "28S40M", true, true, false, false, false, DEFAULT_BASE_QUALITY);
+
```
remove

nh13 commented on src/tests/java/picard/sam/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java:
```diff
@@ -540,4 +549,32 @@ public void testOpticalDuplicateClustersAddingSecondEndFirstSameCoordinate() {
+        tester.addMatePair("RUNID:1:1:15993:13361", 2, 41212324, 41212310, false, false, false, false, "33S35M", "19S49M", true, true, false, false, false, DEFAULT_BASE_QUALITY);
+        tester.addMappedFragment( "RUNID:1:1:15993:13361", 1, 400, markSecondaryAndSupplementaryRecordsLikeTheCannonical(), null, null, additionalFragSecondary, additionalFragSupplementary, DEFAULT_BASE_QUALITY);
+        tester.addMatePair("RUNID:1:1:16020:13352", 2, 41212324, 41212319, false, false, true, true, "33S35M", "28S40M", true, true, false, false, false, DEFAULT_BASE_QUALITY);
+
```
remove

nh13 commented on src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTester.java:
```diff
@@ -0,0 +1,44 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2014 The Broad Institute
```
![image](https://cloud.githubusercontent.com/assets/846274/14057848/ebc51108-f2e2-11e5-82b9-97363c2b617c.png)

nh13 commented on src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTester.java:
```diff
@@ -0,0 +1,44 @@
+public class QuerySortedMarkDuplicatesTester extends AbstractMarkDuplicatesCommandLineProgramTester {
+
+    public QuerySortedMarkDuplicatesTester() {
+
```
remove

nh13 commented on src/tests/java/picard/sam/markduplicates/SimpleMarkDuplicatesWithMateCigar.java:
```diff
@@ -101,7 +102,12 @@ protected int doWork() {
         final SAMFileHeader outputHeader = header.clone();
-        outputHeader.setSortOrder(SAMFileHeader.SortOrder.coordinate);
+
+        if(outputHeader.getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
```
space after `if`

nh13 commented on src/tests/java/picard/sam/markduplicates/SimpleMarkDuplicatesWithMateCigar.java:
```diff
@@ -101,7 +102,12 @@ protected int doWork() {
+        }
+        //outputHeader.setSortOrder(SAMFileHeader.SortOrder.coordinate);
+
         for (final String comment : COMMENT) outputHeader.addComment(comment);
```
opportunity for java 8?

nh13 commented on src/tests/java/picard/sam/testers/SamFileTester.java:
```diff
@@ -23,21 +23,24 @@
+
+
+    public SamFileTester(final int readLength, final boolean deleteOnExit, final int defaultChromosomeLength, final ScoringStrategy duplicateScoringStrategy) {
+     this(readLength, deleteOnExit, defaultChromosomeLength, duplicateScoringStrategy, SAMFileHeader.SortOrder.coordinate);
```
indentation.

nh13 commented on src/tests/java/picard/sam/testers/SamFileTester.java:
```diff
@@ -23,21 +23,24 @@
+        this.samRecordSetBuilder = new SAMRecordSetBuilder(true, sortOrder, true, defaultChromosomeLength, duplicateScoringStrategy);
         samRecordSetBuilder.setReadLength(readLength);
-        setOutputDir();
+        setOutputDir();  }
```
`}` should be on a new line

nh13 commented:
@yfarjoun the changes make sense to me.  I would really like to see it tested on some large datasets too. 99 / :100:

yfarjoun commented:
Thanks for the fast and amusing review. :smile: 

I was planing on testing this on a large bam for speed before merging, but I was anxious to get it "out there" since I couldn't presume such a wonderfully fast review as I got from you. 

I"ll be AFK for a  bit (vacation) and I doubt I'll get to this during the next week, but
I'll post the large data-set test here when I return.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -229,14 +249,30 @@ protected int doWork() {
+
+            if (needNextDuplicateIndex) {
+                    nextDuplicateIndex = (this.duplicateIndexes.hasNext() ? this.duplicateIndexes.next() : NO_SUCH_INDEX);
+                }
```
indentation

yfarjoun commented:
@nh13 I ran a comparison on a 20K read bam and got a 25% increase in run-time (queryname-sorted vs. coordinate-sorted) and identical results (after filtering out the secondary, supplementary, and improperly-mated reads, for which we expect a difference) 

I'm running the same experiment on full-sized bams with different levels of duplication. It will take another few days for the results to come in.  

Also, I don't fully understand the Travis failure...looks like a weird random-number generator issue. I would like to rebase and squash, to see if the problem goes away, is that OK?

nh13 commented:
Squash away

yfarjoun commented:
OK. @nh13: I realized that my tests were inadequate, so I added some more tests, which then required that I change some of the other programs (to collect the new metric that I added). 

I tested this on large files and found no different in the resulting bam (excluding unmapped mates, secondary and supplementary reads) and an increase in runtime of about 25%. I don't think that there's much that can be done about the run-time increase as the sorting collection needs to be shuffled much more when the starting sort-order is query-sorted (twice!) 

I'd love to get a final thumbs up from you.

thanks!

nh13 commented on src/java/picard/sam/DuplicationMetrics.java:
```diff
@@ -41,10 +41,13 @@
 
-    /** The number of mapped read pairs examined. */
+    /** The number of reads that examined that were either Secondary or Supplementary */
+    public long SECONDARY_OR_SUPPLEMENTARY_READS;
```
two things:
1. is this not `_EXAMINED`?  I noticed you changed it back at some point during the many test fixup commits.  Nonetheless it's `UNMAPPED_READS` not `UNMAPPED_READS_EXAMINED` so your call.
2. I would prefer this to go after `READ_PAIRS_EXAMINED`

yfarjoun replied:
1. I dropped the _EXAMINED for two reasons:
- it was too long, 
- I want all the different MarkDuplicates to tally this regardless of how they treat Secondary_orSupplementary reads...so I thought that it would be less confusing that way. 
2. As you wish.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -79,7 +80,12 @@
+                    "The program can take either coordinate-sorted or query-sorted input, however the behavior is slightly different." +
+                    "When the input is coordinate-sorted, unmapped mate of mapped records, and supplementary and secondary alignments are not" +
+                    "marked as duplicates. When the input is query-sorted (actually query-grouped) then unmapped mates get marked as their mapped" +
+                    "counter-parts, and secondary and supplementary reads get marked as the primary records with the same query-name.";
```
this line and the few above need either a newline or a space at the end of each string.

yfarjoun replied:
thanks.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -142,7 +148,10 @@
     private SortingCollection<ReadEndsForMarkDuplicates> fragSort;
     private SortingLongCollection duplicateIndexes;
     private SortingLongCollection opticalDuplicateIndexes;
+    private SAMFileHeader.SortOrder sortOrder;
```
does this need to a member variable or can it be local in `doWork`?

yfarjoun replied:
done.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -142,7 +148,10 @@
+    private SAMFileHeader.SortOrder sortOrder;
+
     private int numDuplicateIndices = 0;
+    private final long NO_SUCH_INDEX = Long.MAX_VALUE; // needs to be large so that that >= test fails for query-sorted traversal
```
probably a good idea to make it static

yfarjoun replied:
done.

nh13 commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -196,10 +205,18 @@ protected int doWork() {
+        log.info("Reads are assumed to be ordered by: " + sortOrder);
+
+        if (sortOrder != SAMFileHeader.SortOrder.coordinate && sortOrder != SAMFileHeader.SortOrder.queryname) {
+            throw new PicardException("This program requires input that are either coordinate or query sorted.");
```
Tell me what they input sort order was too?

nh13 commented on src/java/picard/sam/markduplicates/util/AbstractMarkDuplicatesCommandLineProgram.java:
```diff
@@ -74,12 +73,11 @@
-            doc = "If true, assume that the input file is coordinate sorted even if the header says otherwise.")
-    public boolean ASSUME_SORTED = false;
+            doc = "If not null, assume that the input file has this order even if the header says otherwise.", optional = true)
+    public SAMFileHeader.SortOrder ASSUME_SORT_ORDER = null;
```
I am actually not a fan of this, as this will break everyone's pipeline when they used `ASSUME_SORTED`.  Perhaps keep the old behavior when `ASSUME_SORT_ORDER` is `null` to maintain backwards compatibility?

yfarjoun replied:
what if I also deprecate ASSUME_SORTED?

yfarjoun replied:
meaning, in addition to implementing our suggestion.

nh13 replied:
that works for me.  I am just assuming there are folks who use this option that need some time if they are paying attention to release notes.

nh13 commented on src/tests/java/picard/sam/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java:
```diff
@@ -42,6 +44,13 @@
     protected final static int DEFAULT_BASE_QUALITY = 10;
 
+    protected  boolean markSecondaryAndSupplementaryRecordsLikeTheCannonical(){return false;}
+    protected  boolean markUnmappedRecordsLikeTheirMates(){return false;}
```
spaces

nh13 commented on src/tests/java/picard/sam/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java:
```diff
@@ -42,6 +44,13 @@
+
+    @BeforeClass
+    public void informClass(){
+        System.out.print(this.getClass().getName());
```
Bold move.

yfarjoun replied:
bold...but it doesn't work. taking out.

nh13 commented on src/tests/java/picard/sam/markduplicates/AbstractMarkDuplicatesCommandLineProgramTest.java:
```diff
@@ -42,6 +44,13 @@
+    protected  boolean markUnmappedRecordsLikeTheirMates(){return false;}
+
+    @BeforeClass
+    public void informClass(){
```
`(){` :-1:

nh13 commented on src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java:
```diff
@@ -0,0 +1,46 @@
+ * supplementary and secondary reads will be marked the same as their primary alignments,
+ * and that unmapped reads with mapped mates will be marked according to their mapped mate.
+ */
+
```
`newline` 0 / :100:

nh13 commented on src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java:
```diff
@@ -0,0 +1,46 @@
+    protected boolean markUnmappedRecordsLikeTheirMates() {return true;}
+
+    @Override
+    protected boolean markSecondaryAndSupplementaryRecordsLikeTheCannonical() {return true;}
```
spaces int his file are needed

geoffjentry replied:
Which file is his?

nh13 replied:
I need an `int` # of spaces please.

yfarjoun replied:
I'm lost...what are you asking for here?

nh13 replied:
`{return true;}` -> `{ return true; }`

![image](https://cloud.githubusercontent.com/assets/846274/14515208/93c95760-01c5-11e6-8edd-71a5d51f898c.png)

vdauwera replied:
LOL

If we're going to be pedantic, `Cannonical` -> `Canonical` because cannon != canon. I would illustrate but all google finds for the second term is ![image](https://cloud.githubusercontent.com/assets/2748435/14515383/84c7688e-01f1-11e6-8af8-588eed240823.png) which is not directly relevant. 

For the record, "canonical" [originates in religious law](https://en.wikipedia.org/wiki/Canon_law_(Catholic_Church)), from *jus canonicum* which is itself not directly related to porterhouse steak served with *au jus*. Although delicious, that is also a bit of a misphrasing since *au jus* already includes the *with* bit. So really it's *steak au jus*, period. Still delicious. Still nothing to do with canonical law. No matter.

nh13 commented:
@yfarjoun a few more things, but looking great so far!

yfarjoun commented:
@nh13 I think I addressed your comments. thank-you for your patience, through review, and extreme responsiveness.

nh13 commented:
Extreme! Squash and merge?

yfarjoun commented:
thanks!!

On Thu, Apr 14, 2016 at 10:53 AM, Nils Homer <notifications@github.com>
wrote:

> Extreme! Squash and merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/491#issuecomment-209985097>
>

bradtaylor commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -79,7 +80,12 @@
+                    "                 O=marked_duplicates.bam M=marked_dup_metrics.txt\n" +
+                    "\n" +
+                    "The program can take either coordinate-sorted or query-sorted input, however the behavior is slightly different.\n" +
+                    "When the input is coordinate-sorted, unmapped mate of mapped records, and supplementary and secondary alignments are not\n" +
```
should unmapped mate be unmapped mates?

bradtaylor commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -229,42 +249,65 @@ protected int doWork() {
+                    rec.setDuplicateReadFlag(true);
+
+                    // only update duplicate counts for "decider" reads, not tag-a-long reads
+                    if (!rec.isSecondaryOrSupplementary() && !rec.getReadUnmappedFlag()) {
```
But for queryname sorted runs, 'tagalong' friends are being marked as duplicates right? So why wouldn't we want to update the count to include them? By doing this, will the duplicate metrics not match the count of duplicate tags derived from the file?

yfarjoun replied:
correct, the number will not be the same, but this number is used in order to calculate the library complexity and the ROI calculator, and it would be wrong to use the tagalong reads for this.

bradtaylor commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -229,42 +249,65 @@ protected int doWork() {
 
             if (this.TAGGING_POLICY != DuplicateTaggingPolicy.DontTag && rec.getDuplicateReadFlag()) {
                 if (isOpticalDuplicate) {
+                    duplicateQueryName = rec.getReadName();
```
Should this be opticalDuplicateQueryName?

bradtaylor commented on src/java/picard/sam/markduplicates/MarkDuplicates.java:
```diff
@@ -373,14 +419,22 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
+
             } else if (!rec.isSecondaryOrSupplementary()) {
-                final ReadEndsForMarkDuplicates fragmentEnd = buildReadEnds(header, index, rec, useBarcodes);
+                final long indexForRead = assumedSortOrder == SAMFileHeader.SortOrder.coordinate ? index : duplicateIndex;
```
Would it be clearer to say:
final long indexForRead = assumedSortOrder == SAMFileHeader.SortOrder.queryname ? duplicateIndex : index;
so that it matches up better with the if on 423?

since we only have the two options, they're equivalent, so this is just a readability thing

yfarjoun replied:
sure.

bradtaylor commented on src/tests/java/picard/sam/markduplicates/SimpleMarkDuplicatesWithMateCigar.java:
```diff
@@ -101,8 +102,13 @@ protected int doWork() {
+        if (outputHeader.getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
+            throw new PicardException("This program requires inputs in coordinate SortOrder");
+        }
+        //outputHeader.setSortOrder(SAMFileHeader.SortOrder.coordinate);
```
Remove this instead of commenting it out?

bradtaylor commented:
Looks great! One question about opticalDuplicateQueryName and a couple cosmetic things. Otherwise 👍

yfarjoun commented:
Travis shows that this passed before the rebase.

yfarjoun commented:
thanks all for the reviews! @nh13 @bradtaylor 🙇
